### PR TITLE
Consolidate repository provider

### PR DESCRIPTION
**Pull request by yufeih**
Move repository outside docset

#4259 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5618)

### CODE REVIEW

File: src/docfx/build/document/RepositoryProvider.cs [hunk 1]
```diff
@@ -27,70 +29,93 @@ internal class RepositoryProvider
...
         }
 
+        public (Repository? repository, PathString? pathToRepository) GetRepository(FilePath path)
+        {
+            return path.Origin switch
+            {
+                FileOrigin.Default => GetRepository(Path.Combine(_docsetPath, path.Path)),
+                FileOrigin.Fallback when _localizationProvider != null
+                    => (_localizationProvider.GetFallbackRepositoryWithDocsetEntry().fallbackRepository,
+                        _docsetPathToDefaultRepository is null ? null : _docsetPathToDefaultRepository + path.Path),
```
Comment (live1206) [inline src/docfx/build/document/RepositoryProvider.cs]: can just be `_docsetPathToDefaultRepository`

File: src/docfx/build/document/RepositoryProvider.cs [hunk 2]
```diff
@@ -27,70 +29,93 @@ internal class RepositoryProvider
...
 
-            throw new InvalidOperationException();
+            return (repository, new PathString(Path.GetRelativePath(repository.Path, fullPath)));
         }
 
-        private (string path, Repository?) GetTemplateRepository()
+        private Repository? GetRepositoryCore(string directory)
         {
-            if (_config is null || _packageResolver is null)
+            var repository = GitUtility.FindRepository(directory);
```
Comment (live1206) [inline src/docfx/build/document/RepositoryProvider.cs]: consider rename `repository` to `repoPath`

File: src/docfx/build/document/RepositoryProvider.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/build/document/RepositoryProvider.cs]: This is to get the relative path of a _file_ to repository, used to calculate GitHub edit URL and such, it is different from `GetRepositoryWithDocsetEntry`

File: src/docfx/build/document/RepositoryProvider.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/build/document/RepositoryProvider.cs]: make sense

File: src/docfx/build/document/RepositoryProvider.cs (hunk 1 above)
Comment (live1206) [reply src/docfx/build/document/RepositoryProvider.cs]: I mean `_docsetPathToDefaultRepository is null ? null : _docsetPathToDefaultRepository` can just be `_docsetPathToDefaultRepository`, it is nullable anyway.

Comment (yufeih) [reply src/docfx/build/document/RepositoryProvider.cs]: Now I get it :)

Comment (live1206) [reply src/docfx/build/document/RepositoryProvider.cs]: :)

**Pull request merged**